src/core/api/plug_scanners.go

```diff
@@ -0,0 +1,350 @@
+
+	// Check access permissions
+	if !sa.SecurityCtx.IsAuthenticated() {
+		sa.SendUnAuthorizedError(errors.New("UnAuthorized"))
+		return
+	}
+
+	if !sa.SecurityCtx.IsSysAdmin() {
+		sa.SendForbiddenError(errors.New(sa.SecurityCtx.GetUsername()))
+		return
+	}
+
+	// Use the default controller
+	sa.c = api.DefaultController
+}
+
+// Get the specified scanner
+func (sa *ScannerAPI) Get() {
+	if r := sa.get(); r != nil {
+		// Response to the client
+		sa.Data["json"] = r
+		sa.ServeJSON()
+	}
+}
+
+// List all the scanners
+func (sa *ScannerAPI) List() {
+	p, pz, err := sa.GetPaginationParams()
+	if err != nil {
+		sa.SendBadRequestError(errors.Wrap(err, "scanner API: list all"))
+		return
+	}
+
+	query := &q.Query{
+		PageSize:   pz,
+		PageNumber: p,
+	}
+
+	// Get query key words
+	kws := make(map[string]string)
+	kw := sa.GetString("q")
+	if len(kw) > 0 {
+		ws := strings.Split(kw, ",")
+		for _, w := range ws {
+			tw := strings.TrimSpace(w)
+			kv := strings.Split(tw, "=")
+			if len(kv) == 2 {
+				// valid key and value for the querying keyword
+				kws[kv[0]] = kv[1]
+			}
+		}
+	}
+
+	if len(kws) > 0 {
+		query.Keywords = kws
+	}
+
+	all, err := sa.c.ListRegistrations(query)
+	if err != nil {
+		sa.SendInternalServerError(errors.Wrap(err, "scanner API: list all"))
+		return
+	}
+
+	// Response to the client
+	sa.Data["json"] = all
+	sa.ServeJSON()
+}
+
+// Create a new scanner
+func (sa *ScannerAPI) Create() {
+	r := &scanner.Registration{}
+
+	if err := sa.DecodeJSONReq(r); err != nil {
+		sa.SendBadRequestError(errors.Wrap(err, "scanner API: create"))
+		return
+	}
+
+	if err := r.Validate(false); err != nil {
+		sa.SendBadRequestError(errors.Wrap(err, "scanner API: create"))
+		return
+	}
+
+	// Explicitly check if conflict
+	if !sa.checkDuplicated("name", r.Name) ||
+		!sa.checkDuplicated("url", r.URL) {
+		return
+	}
+
+	uid, err := sa.c.CreateRegistration(r)
+	if err != nil {
+		sa.SendInternalServerError(errors.Wrap(err, "scanner API: create"))
+		return
+	}
+
+	location := fmt.Sprintf("%s/%s", sa.Ctx.Request.RequestURI, uid)
+	sa.Ctx.ResponseWriter.Header().Add("Location", location)
+
+	resp := make(map[string]string, 1)
+	resp["uid"] = uid
+
+	// Response to the client
+	sa.Data["json"] = resp
+	sa.ServeJSON()
+}
+
+// Update a scanner
+func (sa *ScannerAPI) Update() {
+	r := sa.get()
+	if r == nil {
+		// meet error
+		return
+	}
+
+	// full dose updated
+	rr := &scanner.Registration{}
+	if err := sa.DecodeJSONReq(rr); err != nil {
+		sa.SendBadRequestError(errors.Wrap(err, "scanner API: update"))
+		return
+	}
+
+	if err := r.Validate(true); err != nil {
+		sa.SendBadRequestError(errors.Wrap(err, "scanner API: update"))
+		return
+	}
+
+	// Name changed?
+	if r.Name != rr.Name {
+		if !sa.checkDuplicated("name", rr.Name) {
+			return
+		}
+	}
+
+	// URL changed?
+	if r.URL != rr.URL {
+		if !sa.checkDuplicated("url", rr.URL) {
+			return
+		}
+	}
+
+	getChanges(r, rr)
+
+	if err := sa.c.UpdateRegistration(r); err != nil {
```

Comment:
What will happen when update a scanner with the true value of `isDefault`? Maybe there will be more than one default scanners.

Reply:
Default can only be set by the default API, the update API will not pick up that property.

Reply:
It is not possible to update `IsDefault` via the update API.